sdk/test-utils/recorder-v2/src/index.ts

```
@@ -0,0 +1,83 @@
+  private _stopUri: string;
+  private _mode: string;
+
+  constructor(sessionFile: string, playback: boolean) {
+    super();
+    this._sessionFile = sessionFile;
+    this._startUri = playback
+      ? RecordingHttpClient.s_startPlaybackUri
+      : RecordingHttpClient.s_startRecordUri;
+    this._stopUri = playback
+      ? RecordingHttpClient.s_stopPlaybackUri
+      : RecordingHttpClient.s_stopRecordUri;
+    this._mode = playback ? "playback" : "record";
+    this._httpClient = new DefaultHttpClient();
+  }
+
+  async prepareRequest(request: WebResourceLike): Promise<Partial<RequestInit>> {
+    console.log("inside prepare request");
+    await this.start();
+
+    if (!request.headers.contains("x-recording-id")) {
+      request.headers.set("x-recording-id", this._recordingId!);
+      request.headers.set("x-recording-mode", this._mode);
+      const upstreamUrl = url.parse(request.url);
+      const redirectedUrl = { ...upstreamUrl, host: "localhost:5000", protocol: "http" };
+      upstreamUrl.path = undefined;
+      request.headers.set("x-recording-upstream-base-uri", url.format(upstreamUrl));
+      request.url = url.format(redirectedUrl);
+      console.log(redirectedUrl);
+    }
+
+    return await super.prepareRequest(request);
+  }
+
+  async start(): Promise<void> {
+    if (this._recordingId === undefined) {
+      const req = this._createRecordingRequest(this._startUri);
+      const rsp = await this._httpClient.sendRequest(req);
+      if (rsp.status !== 200) {
+        throw new Error("Start request failed.");
+      }
+      const id = rsp.headers.get("x-recording-id");
+      if (!id) {
+        throw new Error("No recording ID returned.");
+      }
+      this._recordingId = id;
+    }
+  }
+
+  async stop(): Promise<void> {
+    if (this._recordingId !== undefined) {
+      const req = this._createRecordingRequest(this._stopUri);
+      req.headers.set("x-recording-save", "true");
+      await this._httpClient.sendRequest(req);
+    }
```

Should we throw an error in the else case? `Stop` before `Start` feels like an error that we should bubble up.

---

Makes sense, adding. :)